### ISSUE TITLE
com.diffplug.spotless.generic.LicenseHeaderStep: Unable to find delimiter regex ^package

### ISSUE THREAD

**Issue by kwin** [diffplug/spotless#1418]
When running `spotless-maven-plugin.2.27.1:apply` on e.g. groovy scripts or Java classes in the unnamed package (sometimes used in tests) it fails with

```
[ERROR] Step 'com.diffplug.spotless.generic.LicenseHeaderStep' found problem in 'src/test/resources/MPLUGIN-314/MyMojo.java':
Unable to find delimiter regex ^package
java.lang.IllegalArgumentException: Unable to find delimiter regex ^package
    at com.diffplug.spotless.generic.LicenseHeaderStep$Runtime.format (LicenseHeaderStep.java:259)
    at com.diffplug.spotless.generic.LicenseHeaderStep$Runtime.access$100 (LicenseHeaderStep.java:194)
    at com.diffplug.spotless.generic.LicenseHeaderStep.lambda$build$7 (LicenseHeaderStep.java:134)
    at com.diffplug.spotless.FormatterFunc.apply (FormatterFunc.java:32)
    at com.diffplug.spotless.FormatterStepImpl$Standard.format (FormatterStepImpl.java:82)
    at com.diffplug.spotless.FormatterStep$Strict.format (FormatterStep.java:88)
    at com.diffplug.spotless.FilterByFileFormatterStep.format (FilterByFileFormatterStep.java:36)
    at com.diffplug.spotless.Formatter.compute (Formatter.java:230)
    at com.diffplug.spotless.PaddedCell.calculateDirtyState (PaddedCell.java:203)
    at com.diffplug.spotless.PaddedCell.calculateDirtyState (PaddedCell.java:190)
    at com.diffplug.spotless.maven.SpotlessApplyMojo.process (SpotlessApplyMojo.java:45)
    at com.diffplug.spotless.maven.AbstractSpotlessMojo.execute (AbstractSpotlessMojo.java:198)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

```

According to https://docs.oracle.com/javase/specs/jls/se19/html/jls-7.html#jls-7.4.2 this is still a valid java file.

**Comment by nedtwigg** [comment]
I can think of a couple workarounds:

- use `targetExclude` to exclude the root-package files
- use your own in-house delimiter like `//nopackage` and then do `licenseHeader('(package )|(//nopackage)', 'APACHEBLAH')`

https://github.com/diffplug/spotless/blob/f312eced28c331bc415aa6b6a4e3fd924b495bfd/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java#L51-L58

But probably all of them are harder than moving your test classes into a package. I'm happy to merge a PR which makes supporting the root package easier, but it simplifies things a whole lot to just require all java code to have package. For example, look at what the scala people are stuck with

- https://github.com/diffplug/spotless/pull/1282

**Comment by blacelle** [comment]
Just hit this myself. Another approach would be to enable skipping `licenseHeader` from  `FormatExceptionPolicyStrict` in `com.diffplug.spotless.maven.FormatterFactory#newFormatter`.

**Comment by blacelle** [comment]
Encountered again. One partial but generic workaround is to go with:

```
<java>
        ....
	<licenseHeader>
		<delimiter>(package )|(import )</delimiter>
	</licenseHeader>
</java>
```

**Comment by blacelle** [comment]
https://github.com/junit-team/junit5/blob/6976f0dd4d5789dfd288ca9183ab05079e4fd98a/gradle/plugins/src/main/kotlin/junitbuild.spotless-conventions.gradle.kts#L30

JUnit5 team went for:

> licenseHeaderFile(license.headerFile, "(package|import|open|module) ")

**Comment by nedtwigg** [comment]
Seems like the JUnit5 workaround is strictly better than our existing value

https://github.com/diffplug/spotless/blob/49e1ab4c178eaad1d654e775639d2c9a14d2c3c6/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java#L53-L55

Unless someone thinks of a reason why we shouldn't change it, I'll change our default to match the above.

**Comment by blacelle** [comment]
Let me add a few tests cases @nedtwigg . Given regex would not match the simplest HelloWorld:

```
public class HelloWorld {
	public static void main(String[] args) {
		System.out.print("Hello World");
	}
}
```

**Comment by blacelle** [comment]
I can not explain the use `open` of the regex. It seems a valid keyword in Kotlin.

I tracked given regex in JUnit5 up to https://github.com/junit-team/junit5/commit/eb2166d3f5c6b68cdf22669d85cce42b2c36ef12#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R338

**Comment by nedtwigg** [comment]
Improvement published in `plugin-gradle 6.17.0` and `plugin-maven 2.35.0` thanks to @blacelle.
